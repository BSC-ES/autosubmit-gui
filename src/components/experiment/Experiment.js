import React, { useEffect, useContext, Fragment } from "react";
import Spinner from "../layout/Spinner";
import { Link } from "react-router-dom";
import ExperimentContext from '../context/experiment/experimentContext';

const Experiment = ({ expidToken, refTree }) => {
  const experimentContext = useContext(ExperimentContext); 
  const { getExperiment, 
        getExperimentGraph, 
        getExperimentTree,
        setAutoUpdatePkl, 
        cleanTreeData,
        loading, 
        loadingState,
        experiment, 
        experimentRunning,
        enabledGraphSearch, 
        data, 
        treedata,
        startAutoUpdatePkl,
        setAutoUpdateRun, 
        startAutoUpdateRun,
        getRunningState} = experimentContext;

  useEffect(() => {
    getExperiment(expidToken);
    getRunningState(expidToken);
    // eslint-disable-next-line
  }, []);
  
  const onSubmitGraph = e => {
    e.preventDefault();
    getExperimentGraph(experiment.expid);
    window.showGraphTab();
  };

  const onSubmitTree = e => {
    e.preventDefault();
    getExperimentTree(experiment.expid);
    // refTree.current.tab('show');
    // var object = this.refs.treeview;
    // object.tab('show');
    window.showTreeTab();
  };

  // const onClearGraph = e => {
  //   e.preventDefault();
  //   cleanGraphData();
  // }

  const onClearTree = e => {
    e.preventDefault();
    cleanTreeData();
  }

  const onJobMonitor = e => {
    e.preventDefault();
    setAutoUpdatePkl(true);
  }

  const onNotJobMonitor = e => {
    e.preventDefault();
    setAutoUpdatePkl(false);

  }

  const onSubmitRun = e => {
    e.preventDefault();
    setAutoUpdateRun(true);
    window.showLogTab();
  };

  const onStopSubmitRun = e => {
    e.preventDefault();
    setAutoUpdateRun(false);
  };

  // const onSubmitGroup = mode => e => {
  //   e.preventDefault();
  //   getExperimentGraphGrouped(experiment.expid, mode);
  // };
  


  // const cleanParentLocal = () => {
  //   console.log('Exp clean ')
  //   cleanParent();
  // };

  const {
    expid,
    owner,
    owner_id,
    path,
    time_last_access,
    time_last_mod,
    description,
    version,
    updateTime,
    error,
    error_message,
    pkl_timestamp,
    model, 
    branch, 
    hpc,
    isGrouped,
  } = experiment;

  if (loading)  return <Spinner />;

  return (
    <div>
      <div className="row">
        <div className="col">
          <Link to='/autosubmitapp/' className='btn btn-light'>
            Back To Search
          </Link>    
        </div>        
      </div>
      <br></br>
      <div className="row">
        <div className="col-12">
        <div className="card">
          <div className='card-header pt-1 pb-0'>
            <div className='row'>
              <div className='col-8 text-left'>
                <h3 className="font-weight-bold">{expid}</h3>
              </div>
              <div className='col-4 text-right'>
                {loadingState && <span className='badge badge-dark text-right'>LOADING...</span>}
                {experimentRunning && !loadingState && <span className='badge badge-success text-right'>RUNNING</span>}
                {!experimentRunning && !loadingState && <span className='badge badge-secondary text-right'>NOT RUNNING</span>}
              </div>
            </div>
          </div>
          <div className='card-text'>
            <div className="row px-3 py-0">
              <div className='col-4'>
                <div><small>{description}</small></div>
                {error && <pre className="text-danger">{error_message}</pre>}
                <div><span>Version: {version}</span></div>
                <div></div>
              </div>

              <div className='col-4'>
                <div><small className='font-weight-bold'>Owner:</small> {owner_id} {owner}</div>
                <div><small className='font-weight-bold'>Path:</small> {path}</div>
                <div><small>SleepTime: {updateTime}</small></div>  
              </div>   
              <div className='col-4'>
                  <div><small className='font-weight-bold'>Last Modified:</small> {time_last_mod}</div>
                  <div><small className='font-weight-bold'>Last Access:</small> {time_last_access}</div>
                  {pkl_timestamp && <Fragment>
                    <div><small className='font-weight-bold'>Pkl:</small> {pkl_timestamp}</div>
                  </Fragment>
                    
                  } 
                
              </div>
            </div>
          </div> 
          <div className="card-footer p-1 text-muted text-center">
            <div className="row">
                <div className="col-md-4">
                    <small><strong>Model:</strong> {model}</small>
                </div>
                <div className="col-md-4">
                    <small><strong>Branch:</strong> {branch}</small>
                </div>
                <div className="col-md-4">
                    <small><strong>Hpc:</strong> {hpc}</small>
                </div>                
            </div>    
          </div>
          <div className='card-footer p-1'>
            <div className='row'>
                <div className='col-md-2'>
                    <form onSubmit={onSubmitGraph} className='form'>
                      <input
                        type='submit'
                        value='Show Graph'
                        className='btn btn-info btn-block btn-sm'
                        disabled={!enabledGraphSearch}
                      />
                    </form>
                </div>
                <div className='col-md-2'>
                  {experiment && !treedata &&
                    <form onSubmit={onSubmitTree} className='form'>
                      <input
                        type='submit'
                        value='Show Tree View'
                        className='btn btn-info btn-block btn-sm'
                        disabled={!enabledGraphSearch}
                      />
                    </form>
                  }
                  {experiment && treedata && 
                    <form onSubmit={onClearTree} className='form'>
                      <input
                        type='submit'
                        value='Clear Tree View'
                        className='btn btn-secondary btn-block btn-sm'
                        disabled={!enabledGraphSearch}
                      />
                    </form>
                  }
                  
                </div>
               
                {/* <div className='col-md-2'>
                  <button className="btn btn-info btn-block btn-sm" data-toggle="modal" data-target="#grouptype" disabled={!enabledGraphSearch}>
                      Group By
                  </button>
                </div>    */}
                <div className="col-md-2">
                    {experiment && !startAutoUpdateRun &&
                      <form onSubmit={onSubmitRun} className='form'>
                        <input
                        type='submit'
                        value='Show Log'
                        className='btn btn-dark btn-sm btn-block'
                        // disabled={!enabledGraphSearch}
                        />
                      </form>
                    }
                    {experiment && startAutoUpdateRun &&
                      <form onSubmit={onStopSubmitRun} className='form'>
                        <input
                        type='submit'
                        value='Hide Log'
                        className='btn btn-secondary btn-sm btn-block'
                        // disabled={!enabledGraphSearch}
                        />
                      </form>
                    }
                </div>       
                <div className='col-md-4'>
                  {experimentRunning && data && !startAutoUpdatePkl && !isGrouped &&
                    <form onSubmit={onJobMonitor} className='form'>
                      <input
                        type='submit'
                        value='Start Job Monitor'
                        className='btn btn-dark btn-block btn-sm'
                        disabled={!enabledGraphSearch}
                      />
                  </form>
                  }
                  {experimentRunning && data && startAutoUpdatePkl &&
                    <form onSubmit={onNotJobMonitor} className='form'>
                      <input
                        type='submit'
                        value='Stop Job Monitor'
                        className='btn btn-danger btn-block btn-sm'
                        disabled={!enabledGraphSearch}
                      />
                  </form>
                  }
                </div>
                
              </div>
          </div>  
          
        </div>
        </div>
      </div>
      {/* <div className="modal fade" id="grouptype" tabIndex="-1" role='dialog' aria-labelledby='grouptypeTitle' aria-hidden='true'>
          <div className="modal-dialog" role="document">
              <div className="modal-content">
                  <div className="modal-header">
                      <h5 className="modal-title" id="grouptypeTitle">
                          Select your grouping option
                      </h5>
                      <button className="close" type="button" data-dismiss='modal' aria-label='Close'>
                          <span aria-hidden='true'>&times;</span>
                      </button>
                  </div>
                  <div className="modal-body">
                    <div className="row">
                      <div className="col-md-3">
                      <form onSubmit={onSubmitGroup('automatic')} className='form'>
                          <input
                            type='submit'
                            value='Automatic'
                            className='btn btn-info btn-block btn-sm'
                            disabled={!enabledGraphSearch}
                          />
                      </form>
                      </div>
                      <div className="col-md-3">
                      <form onSubmit={onSubmitGroup('date')} className='form'>
                          <input
                            type='submit'
                            value='Date'
                            className='btn btn-info btn-block btn-sm'
                            disabled={!enabledGraphSearch}
                          />
                      </form>
                      </div>
                      <div className="col-md-3">
                      <form onSubmit={onSubmitGroup('member')} className='form'>
                          <input
                            type='submit'
                            value='Member'
                            className='btn btn-info btn-block btn-sm'
                            disabled={!enabledGraphSearch}
                          />
                      </form>
                      </div> 
                      <div className="col-md-3">
                      <form onSubmit={onSubmitGroup('chunk')} className='form'>
                          <input
                            type='submit'
                            value='Chunk'
                            className='btn btn-info btn-block btn-sm'
                            disabled={!enabledGraphSearch}
                          />
                      </form>
                      </div>                        
                    </div>
                      
                  </div>
                  <div className="modal-footer">
                      <button type="button" className="btn btn-secondary" data-dismiss="modal">Close</button>
                  </div>
              </div>
          </div>
      </div> */}
    </div>
    
  );
}


export default Experiment;
