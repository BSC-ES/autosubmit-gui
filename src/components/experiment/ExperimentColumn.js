import React, { useContext } from "react";
import Spinner from "../layout/Spinner";
import ExperimentContext from "../context/experiment/experimentContext";

const ExperimentColumn = ({ expidToken, refTree }) => {
  //console.log(expidToken);
  const experimentContext = useContext(ExperimentContext);
  const { loading, experiment } = experimentContext;

  if (experiment === null) return <Spinner />;

  const {
    // expid,
    owner,
    owner_id,
    path,
    //time_last_access,
    time_last_mod,
    version,
    updateTime,
    branch,
    hpc,
    //isGrouped,
  } = experiment;

  if (loading) return <Spinner />;

  return (
    <div>
      <div className='row'>
        <div className='col-12 pr-0'>
          <div className='card'>
            <div className='card-text m-2'>
              <div className='row'>
                <div className='col-md-12'>
                  <div>
                    <small className='font-weight-bold'>Version:</small>
                  </div>
                  <div>{version}</div>
                  <div>
                    <small className='font-weight-bold'>Owner:</small>
                  </div>
                  <div>
                    {owner_id} {owner}
                  </div>
                  <div>
                    <small className='font-weight-bold'>Path:</small>
                  </div>
                  <div>
                    <span>{path}</span>
                  </div>
                  <div>
                    <small className='font-weight-bold'>SleepTime: </small>
                  </div>
                  <div>{updateTime}</div>
                  <div>
                    <small className='font-weight-bold'>Modified:</small>
                  </div>
                  <div>{time_last_mod}</div>
                  {/* <div><small className='font-weight-bold'>Last Access:</small></div>
                    <div>{time_last_access}</div> */}
                </div>
              </div>
            </div>
          </div>
          <div className='card-footer p-1 text-center scroll-x'>
            <div className='row'>
              <div className='col-md-12'>
                <div>
                  <small>Branch:</small>
                </div>
                <div className='lead'>{branch}</div>
                <div>
                  <small>Hpc:</small>
                </div>
                <div className='lead'>{hpc}</div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default ExperimentColumn;
