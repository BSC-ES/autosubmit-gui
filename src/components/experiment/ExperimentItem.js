import React, { useContext } from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
import ExperimentContext from "../context/experiment/experimentContext";

const ExperimentItem = ({
  experiment: { name, description, user, hpc, status, completed, total }
}) => {
  const experimentContext = useContext(ExperimentContext);
  const { getExperimentSummary, summaries } = experimentContext;
  const onGetSummary = e => {
    e.preventDefault();
    console.log(name);
    getExperimentSummary(name);
  };
  return (
    <div className='col mb-4'>
      <div className='card'>
        <div className='card-header text-center py-1'>
          <div className='row'>
            <div className='col-md-3 text-left'>
              <h3 className='font-weight-bold'>{name}</h3>
            </div>
            <div className='col-md-6 text-right'>
              {" "}
              {completed} / {total}{" "}
              <div className='progress'>
                <div
                  className={
                    completed === total
                      ? "progress-bar bg-warning"
                      : status === "RUNNING"
                      ? summaries[name] && summaries[name].n_failed > 0
                        ? "progress-bar progress-bar-striped progress-bar-animated bg-danger"
                        : "progress-bar progress-bar-striped progress-bar-animated bg-success"
                      : summaries[name] && summaries[name].n_failed > 0
                      ? "progress-bar bg-danger"
                      : "progress-bar bg-info"
                  }
                  role='progressbar'
                  style={{
                    width: total > 0 ? (completed / total) * 100 + "%" : "0%"
                  }}
                  aria-valuenow={completed}
                  aria-valuemin='0'
                  aria-valuemax={total}
                ></div>
              </div>
              {/* <span className='badge badge-default'>
              {" "}
              
            </span> */}
            </div>
            <div className='col-md-3 text-right'>
              {status && status === "RUNNING" && (
                <span className='badge badge-success text-right'>{status}</span>
              )}
              {status && status !== "RUNNING" && (
                <span className='badge badge-secondary text-right'>
                  {status}
                </span>
              )}
            </div>
          </div>
        </div>
        <div className='card-body py-1'>
          {/* <h4 className="card-title"></h4> */}
          <h6 className='card-subtitle text-muted pt-2'>
            <small>Owner:</small> {user}
          </h6>
          <p className='card-text py-1 mb-1'>
            <small>{description}</small>
          </p>
          <div className='row'>
            <div className='col-3 px-1'>
              <form onSubmit={onGetSummary} className='form'>
                <input
                  className={
                    summaries[name]
                      ? "btn btn-info btn-block btn-sm"
                      : "btn btn-outline-info btn-block btn-sm"
                  }
                  type='submit'
                  value={summaries[name] ? "Refresh" : "Summary"}
                  aria-controls={name}
                />
              </form>
            </div>
            <div className='col-9 px-1'>
              <Link
                to={`/autosubmitapp/experiment/${name}`}
                className='btn btn-outline-primary btn-block btn-sm'
              >
                More
              </Link>
            </div>
            {summaries[name] && summaries[name].error === false && (
              <div className='col-12 pt-2' id={name}>
                <div className='card card-body p-0'>
                  <div className='row text-center'>
                    <div className='col-md-6'>
                      <small>
                        Avg. Queue {summaries[name].avg_queue_time} min.
                      </small>
                    </div>
                    <div className='col-md-6'>
                      <small>
                        Avg. Run {summaries[name].avg_run_time} min.
                      </small>
                    </div>
                  </div>
                  <div className='row px-2 pb-2'>
                    <div className='col-12'>
                      {summaries[name].n_running > 0 && (
                        <span className='badge badge-success'>
                          Running: {summaries[name].n_running}
                        </span>
                      )}{" "}
                      {summaries[name].n_queuing > 0 && (
                        <span
                          className='badge'
                          style={{ backgroundColor: "pink" }}
                        >
                          Queuing: {summaries[name].n_queuing}
                        </span>
                      )}{" "}
                      {summaries[name].n_submitted > 0 && (
                        <span
                          className='badge'
                          style={{ backgroundColor: "cyan" }}
                        >
                          Submitted: {summaries[name].n_submitted}
                        </span>
                      )}{" "}
                      {summaries[name].n_suspended > 0 && (
                        <span
                          className='badge'
                          style={{ backgroundColor: "orange" }}
                        >
                          Suspended: {summaries[name].n_suspended}
                        </span>
                      )}
                    </div>
                  </div>

                  {summaries[name].n_failed > 0 && (
                    <div className='row px-1'>
                      <div className='col-md-3 text-center px-2'>
                        <span className='badge badge-danger'>
                          Failed: {summaries[name].n_failed}
                        </span>
                      </div>
                      <div className='col-md-9 text-left'>
                        <div
                          className=''
                          style={{ overflow: "auto", maxHeight: "200px" }}
                        >
                          {summaries[name].failed_jobs.map(item => (
                            <li key={item}>
                              <small>{item}</small>
                            </li>
                          ))}
                        </div>
                      </div>
                    </div>
                  )}
                </div>
              </div>
            )}
          </div>
          <p className='card-text text-center'>
            <small className='text-muted'>HPC: {hpc}</small>
          </p>
        </div>
      </div>
    </div>
  );
};

ExperimentItem.propTypes = {
  experiment: PropTypes.object.isRequired
};

export default ExperimentItem;
