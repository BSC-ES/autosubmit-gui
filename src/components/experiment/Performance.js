import React, { useContext, Fragment } from "react";
import ExperimentContext from "../context/experiment/experimentContext";

const Performance = () => {
  const experimentContext = useContext(ExperimentContext);
  const { performancedata, experiment } = experimentContext;

  if (!experiment || !performancedata) {
    return (
      <div className='row'>
        <div className='col-12'>
          <div className='lead'>Experiment not defined.</div>
        </div>
      </div>
    );
  }
  console.log(performancedata);
  const { ASYPD, SYPD, Parallelization, CHSY, considered } = performancedata;
  return (
    <Fragment>
      <div className='row px-3'>
        <div className='col-6 p-4'>
          <h5>
            Parallelization:{" "}
            <span className='badge badge-secondary'>{Parallelization}</span>
          </h5>
          <h5>
            SYPD: <span className='badge badge-secondary'>{SYPD}</span>
          </h5>
          <h5>
            ASYPD: <span className='badge badge-secondary'>{ASYPD}</span>
          </h5>
          <h5>
            CHSY: <span className='badge badge-secondary'>{CHSY}</span>
          </h5>
          {/* <strong>considered: </strong> {considered} */}
        </div>
        <div className='col-6 p-3 scroll-y-jobs'>
          <p className='lead'>Considered: ({considered.length})</p>
          <ol>
            {considered
              .sort((a, b) => (a.name > b.name ? 1 : -1))
              .map((item) => (
                <li key={item.name}>
                  {" "}
                  {item.name} | QUEUE: <strong>{item.queue}</strong> | RUNNING:{" "}
                  <strong>{item.running}</strong> | CHSY:{" "}
                  <strong>{item.CHSY}</strong>{" "}
                </li>
              ))}
          </ol>
        </div>
      </div>
      <div className='row px-3'>
        <div className='col-12 px-4'>
          <p className='lead'>Metrics description:</p>
          <p>
            <strong>Parallelization</strong>: Total number of cores allocated
            for the run, per SIM.
          </p>
          <p>
            <strong>SYPD</strong>: Simulated years per day for the model in a 24
            h period.
          </p>
          <p>
            <strong>ASYPD</strong>: Actual SYPD: This number should be lower
            than SYPD due to interruptions, queue wait time, data transfer or
            issues with the model workflow. This is collected by measuring the
            time between first submission and the date of arrival of the last
            history file on the storage file system.
          </p>
          <p>
            <strong>CHSY</strong>: Core hours per simulated year. This is
            measured as the product of the model runtime for 1 SY and the number
            of cores allocated. This is an average of the CHSY of all SIM jobs.
          </p>
          <p>
            <strong>Considered</strong>: Scrollable list where each item in the
            list is a job that shows <i>job name</i>, <i>QUEUE time</i> in
            seconds, <i>RUNNING time</i> in seconds, and <i>CHSY</i> for that
            job.
          </p>
          <p>
            Visit{" "}
            <a href='https://earth.bsc.es/gitlab/wuruchi/autosubmitreact/-/wikis/Performance-Metrics'>
              Performance Metrics Documentation
            </a>{" "}
            for more details.
          </p>
        </div>
      </div>
    </Fragment>
  );
};

export default Performance;
