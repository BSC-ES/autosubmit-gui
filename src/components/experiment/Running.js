import React, { Component, Fragment } from 'react'

class Running extends Component {
    constructor(props) {
        super(props);
        this.rundata = null;
    }
    // componentDidUpdate() {
    //     //console.log(this.props.experiment.updateTime)
    //     // if(this.props.autoUpdateLog) {
    //     //     console.log(this.props.autoUpdateLog)
    //     //     this.interval = setInterval(() => this.props.getExperimentRun(this.props.experiment.expid) , this.props.experiment.updateTime);
    //     // }        
    // }

    componentDidMount(){
        this.props.getExperimentRun(this.props.experiment.expid);
        this.interval = setInterval(() => this.props.getExperimentRun(this.props.experiment.expid) , this.props.experiment.updateTime * 1000);
        // if (this.props.rundata){
        //     console.log('DidMount Running.')
        // }
    }

    componentWillUnmount() {
        this.props.cleanRunData();
        clearInterval(this.interval);
    }
    
    
    render() {

        if (this.props.rundata === null){
            return (
              <Fragment>
                <div>Loading...</div>
              </Fragment>
            );       
          }

        const { logcontent,
        error,
        error_message,
        found,
        logfile,
        lastModified,
        timeStamp} = this.props.rundata;

        const logitems = logcontent.map((item) =>
            <li key={item.index}><small>{item.content}</small></li>
        );
        
        if (error) {
            return (
                <Fragment>
                    <div>{error_message}</div>
                    <br></br>
                    <hr></hr>
                </Fragment>
                
            );
        }
        if (!found) {
            return (
                <Fragment>
                    <div>Running log not found. This experiment might not be currently running on Autosubmit.</div>
                    <br></br>
                    <hr></hr>
                </Fragment>                
            );
        }

        return (
            <div>
                <h3>Logfile: {logfile}</h3>
                <p>Last Modified: {lastModified}</p>
                <p><small>TimeStamp: {timeStamp}</small></p>
                <p><strong>Showing last 40 lines of the log file:</strong></p>
                <pre className="bash">
                    <ul style={pStyle}>{logitems}</ul>                    
                </pre>
                <br></br>
                <h3>End of log file.</h3>
                <br></br>
            </div>
        )
    }
}

const pStyle = {
    'listStyleType': 'none',
};


export default Running;