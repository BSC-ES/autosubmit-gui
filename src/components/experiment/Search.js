import React, { useState, useContext, Fragment } from 'react';
import ExperimentContext from '../context/experiment/experimentContext';
import AlertContext from '../context/alert/alertContext';

const Search = ({ setAlert }) => {
  const experimentContext = useContext(ExperimentContext);
  const alertContext = useContext(AlertContext);

  const [text, setText] = useState('');

  const onSubmit = e => {
    e.preventDefault();
    if (text === '') {
      alertContext.setAlert('Please enter something', 'light');
    } else {
      experimentContext.searchExperiments(text);
      //setText('');
    }
  };

  const onChange = e => setText(e.target.value);

  return (
    <div className="container">
      <form onSubmit={onSubmit} className='form'>
        <div className="input-group mb-3">
          
            <input
              className="form-control"
              type='text'
              name='text'
              placeholder='Search Experiments by Expid or Description...'
              value={text}
              onChange={onChange}
            />
            <div className="input-group-append">
              <input          
                type='submit'
                value='Search'
                className='btn btn-dark'
              />
            </div>        
        </div>
      </form>
      {experimentContext.experiments.length > 0 && (
            <Fragment>
              <button className='btn btn-light btn-block' onClick={experimentContext.clearExperiments}>
                Clear
              </button>
              <br></br>
            </Fragment>                                  
          )} 
    </div>
  );
}

export default Search;
