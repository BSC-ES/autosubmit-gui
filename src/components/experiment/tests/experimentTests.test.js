import React from "react";
import { unmountComponentAtNode, render } from "react-dom";
import { BrowserRouter as Router } from "react-router-dom";
import renderer from 'react-test-renderer';
import { act } from "react-dom/test-utils";
import ExperimentItem from "../ExperimentItem";
import Search from "../Search";
import ExperimentColumn from "../ExperimentColumn";
import Experiment from "../Experiment";
import Performance from "../Performance";
import Running from "../Running";
import ExperimentContext from "../../context/experiment/experimentContext";
import TreeContext from "../../context/tree/treeContext";
import AlertContext from "../../context/alert/alertContext";

let container = null;
beforeEach(() => {
  // setup DOM element as render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

// --- ExperimentItem.js ---

it("ExperimentItem renders with content", () => {
  const experiment = {name: "a2h6", description: "Test", user:"wuruchi", hpc:"LOCAL", status:"RUNNING", completed:10, total:100, version:"3.11.0"};
  act(() => {
    render(<Router><ExperimentContext.Provider value={{ experiment: {}, summaries: new Map(), loadingSummary: new Map()}}><ExperimentItem key="a2h6" experiment={experiment} summaries={new Map()} loadingSummary={new Map()}/></ExperimentContext.Provider></Router>, container);
  });
  expect(container.textContent).toContain("a2h6");
  expect(container.textContent).toContain("wuruchi");
});


it('ExperimentItem renders correctly', () => {
  const experiment = {name: "a2h6", description: "Test", user:"wuruchi", hpc:"LOCAL", status:"RUNNING", completed:10, total:100, version:"3.11.0"};
  const tree = renderer.create(<Router><ExperimentContext.Provider value={{ experiment: {}, summaries: new Map(), loadingSummary: new Map()}}><ExperimentItem key="a2h6" experiment={experiment} summaries={new Map()} loadingSummary={new Map()}/></ExperimentContext.Provider></Router>).toJSON();
  expect(tree).toMatchSnapshot();
});



// --- ExperimentColumns.js --- 

it('ExperimentColumn renders with content', () => {
  const experiment = {branch: "master", description: "Test", owner:"wuruchi", owner_id: 1226, hpc:"LOCAL", db_historic_version: 14, version:"3.11.0"};
  act(() => {
    render(<Router><ExperimentContext.Provider value={{ experiment: experiment, totalJobs: 100, loading: false}}><ExperimentColumn /></ExperimentContext.Provider></Router>, container);
  });
  expect(container.textContent).toContain("master");
  expect(container.textContent).toContain("wuruchi");
  expect(container.textContent).toContain("100");
  expect(container.textContent).toContain("14");
});

// --- Experiment.js --- 

it('Experiment renders with content', () => {

  const experiment = {expid: "a2h6", branch: "master", description: "Test", owner:"wuruchi", owner_id: 1226, hpc:"LOCAL", db_historic_version: 14, version:"3.11.0"};

  act(() => {
    render(<Router><ExperimentContext.Provider value={{ experiment: experiment, totalJobs: 100, loading: false, loadingState: false, experimentRunning: true}}><TreeContext.Provider value={{ getExperimentRunJobData: () => null, fancyTree: {}, startAutoUpdateTreePkl: false, loadingPreviousRun: false, currentRunIdOnTree: {runId: 1, message: "None", created: "2020-11-02-16:59:42"} }}>
          <Experiment />
        </TreeContext.Provider>        
      </ExperimentContext.Provider>
      </Router>, container);
  });
  expect(container.textContent).toContain("a2h6");
  expect(container.textContent).toContain("ACTIVE");

  act(() => {
    render(<Router><ExperimentContext.Provider value={{ experiment: experiment, totalJobs: 100, loading: false, loadingState: false, experimentRunning: false}}>
    <TreeContext.Provider value={{ getExperimentRunJobData: () => null, fancyTree: null, startAutoUpdateTreePkl: false,  loadingPreviousRun: false, currentRunIdOnTree: {runId: 1, message: "None", created: "2020-11-02-16:59:42"} }}>
    <Experiment />
    </TreeContext.Provider>      
    </ExperimentContext.Provider></Router>, container);
  });
  expect(container.textContent).toContain("a2h6");
  expect(container.textContent).toContain("INACTIVE");
});

// --- Running.js ---

it('Log Running renders with no rundata', () => {
  const experiment = {expid: "a2h6"};
  //const rundata
  act(() => {
    render(<Running rundata={null}
      loadingRun={false}
      cleanRunData={() => null}
      getExperimentRun={() => null}
      experiment={experiment}
      startAutoUpdateRun={() => null}
      setAutoUpdateRun={() => null}
      experimentRunning={false} />, container);
  });
  expect(container.textContent).toContain("Loading");
})

it('Log Running renders with rundata', () => {
  const experiment = {expid: "a2h6"};
  const rundata = {error: false, erro_message: "", found: true, lastModified: "2020-11-02 16:59:42", logcontent: [
    {
        content: "  File \"/home/Earth/wuruchi/AutosubmitRealDeal/autosubmit-issue371a/autosubmit/autosubmit/database/db_jobdata.py\", line 1321, in process_current_run_collection",
        index: 0
    },
    {
      content: "    if jobs_in_package[i].running_time() <= 0:",
      index: 1
    },
    {
      content: "IndexError: list index out of range",
      index: 2
    }],
  logfile: "20201102_175754_run.log",
  timeStamp: 1604336382.4040346}
  act(() => {
    render(<Running rundata={rundata}
      loadingRun={false}
      cleanRunData={() => null}
      getExperimentRun={() => null}
      experiment={experiment}
      startAutoUpdateRun={() => null}
      setAutoUpdateRun={() => null}
      experimentRunning={false} />, container);
  });
  expect(container.textContent).toContain("IndexError: list index out of range");
  expect(container.textContent).toContain("20201102_175754_run.log");
});
