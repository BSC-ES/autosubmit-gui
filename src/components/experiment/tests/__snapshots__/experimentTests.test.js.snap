// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExperimentItem renders correctly 1`] = `
<div
  className="col mb-4"
>
  <div
    className="card"
  >
    <div
      className="card-header text-center py-1"
    >
      <div
        className="row"
      >
        <div
          className="col-md-3 text-left"
        >
          <h3
            className="font-weight-bold"
          >
            a2h6
          </h3>
        </div>
        <div
          className="col-md-6 text-center"
        >
           
          10
           / 
          100
           
          <div
            className="progress border"
          >
            <div
              aria-valuemax={100}
              aria-valuemin="0"
              aria-valuenow={10}
              className="progress-bar progress-bar-striped progress-bar-animated bg-success"
              role="progressbar"
              style={
                Object {
                  "width": "10%",
                }
              }
            />
          </div>
        </div>
        <div
          className="col-md-3 text-right"
        >
          <span
            className="badge badge-success text-right"
          >
            ACTIVE
          </span>
        </div>
      </div>
    </div>
    <div
      className="card-body py-1"
    >
      <div
        className="d-flex justify-content-between"
      >
        <div>
          <h6
            className="card-subtitle text-muted pt-2"
          >
            <span>
              Owner:
            </span>
             
            wuruchi
          </h6>
        </div>
        <div>
          <span
            className="text-muted"
          >
            HPC: 
            LOCAL
          </span>
        </div>
      </div>
      <p
        className="card-text py-1 mb-1"
      >
        <span>
          Test
        </span>
      </p>
      <div
        className="row"
      >
        <div
          className="col-3 px-1"
        >
          <form
            className="form"
            onSubmit={[Function]}
          >
            <input
              aria-controls="a2h6"
              className="btn btn-primary btn-block btn-sm"
              data-placement="bottom"
              data-toggle="tooltip"
              title="Shows a summary of the current progress of the experiment."
              type="submit"
              value="Summary"
            />
          </form>
        </div>
        <div
          className="col-3 px-1"
        >
          <a
            className="btn btn-primary btn-block btn-sm"
            data-placement="bottom"
            data-toggle="tooltip"
            href="/autosubmitapp/experiment/a2h6"
            onClick={[Function]}
            title="Opens the experiment page where the Tree View representation is loaded by default."
          >
            Tree
          </a>
        </div>
        <div
          className="col-3 px-1"
        >
          <a
            className="btn btn-primary btn-block btn-sm"
            data-placement="bottom"
            data-toggle="tooltip"
            href="/autosubmitapp/experiment/a2h6/graph"
            onClick={[Function]}
            title="Opens the experiment page where the Graph View representation is loaded by default."
          >
            Graph
          </a>
        </div>
        <div
          className="col-3 px-1"
        >
          <a
            className="btn btn-primary btn-block btn-sm"
            data-placement="bottom"
            data-toggle="tooltip"
            href="/autosubmitapp/experiment/a2h6/light"
            onClick={[Function]}
            title="Opens the experiment page where a simple list of jobs and their status is presented. Loads quicker than the Tree View."
          >
            Quick
          </a>
        </div>
      </div>
      <p
        className="card-text text-center"
      >
        <span
          className="text-muted"
        >
          3.11.0
        </span>
      </p>
    </div>
  </div>
</div>
`;
